### PR TITLE
Methods discriminator model

### PR DESCRIPTION
**Pull request by danich1**
This PR describes the discriminator model and how we calibrated the model to emit reliable probability estimates. The svgs are the image files I used to create the network figure.

### CODE REVIEW

File: content/03.methods.md [hunk 1]
```diff
@@ -14,9 +14,35 @@ talk about generative model and how it works
 ### Word Embeddings
 mention facebooks fasttext model and how we used it to train word vectors
 ### Discriminator Model
-talk about the discriminator model and how it works
+The discriminator model is designed to improve off of the predctions from the generative model.
```
Comment (dhimmel) [inline content/03.methods.md]: "predctions" misspelling. Is it confusing to call the generative output "predictions"? Should we reserve the word predictions for the discriminator output? Let's try to find a consistent terminology for the output value from the generative model.

File: content/03.methods.md [hunk 2]
```diff
@@ -14,9 +14,35 @@ talk about generative model and how it works
...
+We used a convolutional neural network with multiple filters as our discriminator model.
+This network uses multiple filters with fixed widths of 300 dimensions and can have varying heights (Figure 1).
+For our model we each filter at a fixed height of 7, because this provided the best performance in terms of relationship classification [@arxiv:1510.03820].
+We trained this model for 20 epochs using the adam optimizer [@arxiv:1412.6980] with a learning rate of 0.001.
+This optimizer had beta values of 0.9 and 0.99 respectively.
+We added a L2 penalty on the network weights to prevent overfitting.
+Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.
+
+![convolutional neural network here](images/figures/convolutional_neural_network/convolutional_neural_nework.png)
```
Comment (dhimmel) [inline content/03.methods.md]: Figure caption syntax

**Review by dhimmel — COMMENTED** [review]
Seems that there are many figures that are just components of the final figure. Best to not put these in `content/images` because this causes much duplication. Perhaps make a `figures` directory in the root where you keep the source files used to generate the figures.

File: content/03.methods.md (hunk 1 above)
Comment (cgreene) [reply content/03.methods.md]: How about "generated labels"?

File: content/images/figures/convolutional_neural_network/kernel filters.svg [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
```
Comment (cgreene) [inline content/images/figures/convolutional_neural_network/kernel filters.svg]: from my own past experience, i find spaces in filenames (or other characters that need to be escaped) are something to always avoid

File: content/03.methods.md [hunk 4]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
 ### Word Embeddings
 mention facebooks fasttext model and how we used it to train word vectors
 ### Discriminator Model
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
### Discriminative Model
```
Making the construction between generative/discriminative parallel

File: content/03.methods.md [hunk 5]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
-talk about the discriminator model and how it works
+The discriminator model is designed to improve off of the generated labels from the generative model.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
The discriminative model is a neural network, which we train to predict labels from the generative model.
The expectation is that the discriminative model can learn more complete features of the text than the label functions that are used in the generative model.
```

File: content/03.methods.md [hunk 6]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+We used a convolutional neural network with multiple filters as our discriminator model.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
We used a convolutional neural network with multiple filters as our discriminative model.
```

File: content/03.methods.md [hunk 7]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+This network uses multiple filters with fixed widths of 300 dimensions and can have varying heights (Figure {@fig:convolutional_network}).
```
Comment (cgreene) [inline content/03.methods.md]: I don't quite understand this sentence. Is it about CNNs in general, or your specific one?

File: content/03.methods.md [hunk 8]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+For our model we each filter at a fixed height of 7, because this provided the best performance in terms of relationship classification [@arxiv:1510.03820].
```
Comment (cgreene) [inline content/03.methods.md]: Is this based on the reference from the arxiv paper?

File: content/03.methods.md [hunk 9]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+We trained this model for 20 epochs using the adam optimizer [@arxiv:1412.6980] with a learning rate of 0.001.
+This optimizer had beta values of 0.9 and 0.99 respectively.
```
Comment (cgreene) [inline content/03.methods.md]: Why respectively? I might have missed that there are two things there.

File: content/03.methods.md [hunk 10]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+We added a L2 penalty on the network weights to prevent overfitting.
+Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.
+
+![
+This is the architecture of our convolutional neural network. We treat the sentence matrix like an image and perform a convolution step using multiple filters.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters. 
```

File: content/03.methods.md [hunk 11]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
+![
+This is the architecture of our convolutional neural network. We treat the sentence matrix like an image and perform a convolution step using multiple filters. 
+These filters generate a feature map that is sent into a maximum pooling layer. 
+This layer extracts the largest feature in each of these maps. The extract features are concatenated into a singular vector that will be passed into a fully connected network. 
+The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 
+From the fully connected network the last step is to generate predictions using the softmax layer.
+](images/figures/convolutional_neural_network/convolutional_neural_nework.png){#fig:convolutional_network}
+
+
 ### Discriminator Model Calibration
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
### Calibration of the Discriminative Model
```

File: content/03.methods.md [hunk 12]
```diff
@@ -14,9 +14,36 @@ talk about generative model and how it works
...
-talk about calibrating deep learning models with temperature smoothing
+Often many tasks require a machine learning model to output reliable probability predictions. 
+This means if a model were to assign a class label with 80% probability, that class should appear 80% of the time.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
A model is well calibrated if the probabilities emitted from the model match the observed probabilities: a well-calibrated model that assigns a class label with 80% probability should have that class appear 80% of the time.
```

**Review by cgreene — CHANGES_REQUESTED** [review]
A few more thoughts! Looking good :)

File: content/03.methods.md (hunk 8 above)
Comment (danich1) [reply content/03.methods.md]: Yeah this was based off of the arxiv paper. They did bench marking experiments using the same model I'm using for this project.

File: content/03.methods.md (hunk 7 above)
Comment (danich1) [reply content/03.methods.md]: This line is about my specific model.

File: content/03.methods.md (hunk 9 above)
Comment (danich1) [reply content/03.methods.md]: The adam optimizer takes two values for beta. According to the [pytorch documentation](https://pytorch.org/docs/stable/optim.html#torch.optim.Adam):
> **beta** (Tuple[float, float], optional) - coefficients used for computing running averages of gradient and its square

Turns out the values I'm using are the default, so I can change this to say default settings if too confusing.

Comment (cgreene) [reply content/03.methods.md]: Yes - I think that would be helpful

File: content/03.methods.md (hunk 7 above)
Comment (cgreene) [reply content/03.methods.md]: Here it says the heights can vary, but below it says 7. Should it just say a height of 7 `because this provided...`?

File: content/03.methods.md [hunk 13]
```diff
@@ -13,10 +13,38 @@ describe what a label function is and how many we created for each relation
...
+We used a convolutional neural network with multiple filters as our discriminative model.
+This network uses multiple filters with fixed widths of 300 dimensions and can have varying heights (Figure {@fig:convolutional_network}).
+For our model we each filter at a fixed height of 7, because this provided the best performance in terms of relationship classification [@arxiv:1510.03820].
+We trained this model for 20 epochs using the adam optimizer [@arxiv:1412.6980] with a learning rate of 0.001.
+This optimizer used pytorch's default parameter settings.
+We added a L2 penalty on the network weights to prevent overfitting.
+Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.
+
+![
+The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
The architecture of the discriminative model is a convolutional neural network.
We perform a convolution step using multiple filters. 
```

File: content/03.methods.md [hunk 14]
```diff
@@ -13,10 +13,38 @@ describe what a label function is and how many we created for each relation
...
+For our model we each filter at a fixed height of 7, because this provided the best performance in terms of relationship classification [@arxiv:1510.03820].
+We trained this model for 20 epochs using the adam optimizer [@arxiv:1412.6980] with a learning rate of 0.001.
+This optimizer used pytorch's default parameter settings.
+We added a L2 penalty on the network weights to prevent overfitting.
+Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.
+
+![
+The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters. 
+These filters generate a feature map that is sent into a maximum pooling layer. 
+This layer extracts the largest feature in each of these maps. The extract features are concatenated into a singular vector that will be passed into a fully connected network.
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
This layer extracts the largest feature in each of these maps.
The extracted features are concatenated into a singular vector that is passed into a fully connected network. 
```

File: content/03.methods.md [hunk 15]
```diff
@@ -13,10 +13,38 @@ describe what a label function is and how many we created for each relation
...
+We added a L2 penalty on the network weights to prevent overfitting.
+Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.
+
+![
+The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters. 
+These filters generate a feature map that is sent into a maximum pooling layer. 
+This layer extracts the largest feature in each of these maps. The extract features are concatenated into a singular vector that will be passed into a fully connected network. 
+The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 
+From the fully connected network the last step is to generate predictions using the softmax layer.
+](images/figures/convolutional_neural_network/convolutional_neural_nework.png){#fig:convolutional_network}
```
Comment (cgreene) [inline content/03.methods.md:174]: This figure needs some improvement. It looks like there are some parts that are close-to-but-not-quite aligned. They should either be aligned, or they should be clearly and intentionally unaligned. There's also very different font sizes, styles, and maybe even widths.

File: content/03.methods.md (hunk 7 above)
Comment (danich1) [reply content/03.methods.md]: Yeah I think saying height of 7 is appropriate here. Will change.

**Comment by cgreene** [comment]
Screen Shot 2019-07-23 at 10 34 23 AM

This is what I see for the image. It still looks like it's not aligned. 

It will probably be easier to use these options to get things to line up.
Screen Shot 2019-07-23 at 10 35 53 AM

Screen Shot 2019-07-23 at 10 36 05 AM

**Review by cgreene — APPROVED** [review]

File: content/03.methods.md [hunk 16]
```diff
@@ -1,22 +1,200 @@
...
+The extracted features are concatenated into a singular vector that is passed into a fully connected network. 
+The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 
+From the fully connected network the last step is to generate predictions using the softmax layer.
+](images/figures/convolutional_neural_network/convolutional_neural_nework.png){#fig:convolutional_network}
+
+#### Calibration of the Discriminative Model
+
+Often many tasks require a machine learning model to output reliable probability predictions. 
+A model is well calibrated if the probabilities emitted from the model match the observed probabilities: a well-calibrated model that assigns a class label with 80% probability should have that class appear 80% of the time.
+Typically, deep learning models tend to be poorly calibrated [@arxiv:1706.04599; @arxiv:1807.00263].
```
Comment (cgreene) [inline content/03.methods.md]:
```suggestion
Deep neural network models can often be poorly calibrated [@arxiv:1706.04599; @arxiv:1807.00263].
```
